File: src/other/PathSum.py
from typing import Optional

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:
        if root is None:
            return False
        elif root.left is None and root.right is None and targetSum == root.val:
            return True
        else:
            return self.hasPathSum(root.left, targetSum - root.val) or self.hasPathSum(root.right, targetSum - root.val)
        
s = Solution()
tree = TreeNode(5, TreeNode(4, TreeNode(11, TreeNode(7), TreeNode(2))), TreeNode(8, TreeNode(13), TreeNode(4, None, TreeNode(1))))
print(s.hasPathSum(tree, 22))

File: src/leetcode/PathSum.java
package Programming.leetcode;

import Programming.leetcode.util.TreeNode;

public class PathSum {
  public boolean hasPathSum(TreeNode root, int targetSum) {
    if (root == null) return false;
    else if (root.left == null && root.right == null) return root.val == targetSum;
    else if (root.right == null) return hasPathSum(root.left, targetSum - root.val);
    else if (root.left == null) return hasPathSum(root.right, targetSum - root.val);
    else return hasPathSum(root.left, targetSum - root.val)
              || hasPathSum(root.right, targetSum - root.val);
  }
}
